src/_pytest/python_api.py

```diff
@@ -668,30 +707,50 @@ def raises(expected_exception, *args, match=None, **kwargs):
         try:
             func(*args[1:], **kwargs)
         except expected_exception:
-            return _pytest._code.ExceptionInfo.from_current()
+            # Cast to narrow the type to expected_exception (_E).
+            return cast(
```

Comment:
hmmm this cast is a little unfortunate -- should we maybe ditch `from_current` here and perform the full construction using `exc_info` (for better type-safety?)

Reply:
`from_current` does some stuff I don't understand, using the constructor directly would bypass that, but we can add a new classmethod to construct from an existing `exc_info`. I'll try it.